source/extensions/filters/http/transformation/inja_transformer.cc

```diff
@@ -105,6 +132,77 @@ Extractor::extractValue(Http::StreamFilterCallbacks &callbacks,
   return "";
 }
 
+// Match a regex against the input value and replace the matched subgroup with the replacement_text_ value
+// writes the result to replaced_value_ and returns a absl::string_view to it
+absl::string_view
+Extractor::replaceIndividualValue(Http::StreamFilterCallbacks &callbacks,
+                                  absl::string_view value) const {
+  std::match_results<absl::string_view::const_iterator> regex_result;
+
+  // if there are no matches, return an empty string
+  if (!std::regex_search(value.begin(), value.end(), regex_result, extract_regex_)) {
+    ENVOY_STREAM_LOG(debug, "replaceValue: extractor regex did not match input", callbacks);
+    return "";
+  }
+
+  // if the subgroup specified is greater than the number of subgroups in the regex, return an empty string
+  if (group_ >= regex_result.size()) {
+    ASSERT("no such group in the regex");
+    ENVOY_STREAM_LOG(debug, "replaceValue: invalid group specified for regex", callbacks);
+    return "";
+  }
+
+  // if the regex doesn't match the entire input value, return an empty string
+  if (regex_result[0].length() != long(value.length())) {
+    ENVOY_STREAM_LOG(debug, "replaceValue: Regex did not match entire input value. This is not allowed when replace_all is false", callbacks);
+    return "";
+  }
+
+  // Create a new string with the input value
+  std::string replaced(value.begin(), value.end());
+
+  // Replace the specified subgroup with the replacement_text_ value
+  const auto &sub_match = regex_result[group_];
+  auto start_pos = sub_match.first - value.begin();
+  auto length = sub_match.length();
+  replaced.replace(start_pos, length, replacement_text_);
+
+  // Store the replaced string in replaced_value_ and return an absl::string_view to it
+  replaced_value_ = std::move(replaced);
+  return absl::string_view(replaced_value_);
+}
+
+// Match a regex against the input value and replace all instances of the regex with the replacement_text_ value
+// writes the result to replaced_value_ and returns a absl::string_view to it
+absl::string_view
+Extractor::replaceAllValues(Http::StreamFilterCallbacks &callbacks,
+                            absl::string_view value) const {
+  std::string input(value.begin(), value.end());
+  std::string replaced;
+  bool matchFound = false;
+
+  // create an iterator to search for matches of extract_regex_ in the input string
+  std::sregex_iterator it(input.begin(), input.end(), extract_regex_);
+  std::sregex_iterator end_it; // default end iterator for comparison
+
+  // set matchFound to true if at least one match is found
+  if (it != end_it) {
+    matchFound = true; 
+  }
+
+  if (!matchFound) {
+    ENVOY_STREAM_LOG(debug, "replaceAllValues: extractor regex did not match input", callbacks);
+    return "";
+  }
+
+  // If a match was found, replace all instances of the regex in the input value with the replacement_text_ value
+  replaced = std::regex_replace(input, extract_regex_, replacement_text_);
```

Comment:
similar performance concerns here since we have had issues with `std::regex` in the past

Reply:
Some notes/questions:
 - Extractor regexes are compiled at config time, which is good (or really, the alternative is unacceptable)
   - https://github.com/solo-io/envoy-gloo/blob/7da296e3c690b02aa0fe1bc879950cdf0ea40ee9/source/extensions/filters/http/transformation/inja_transformer.cc#L59
 - The replaceAll path (i.e. the one we're discussing here) is not maximally efficient -- it does an efficient check to identify whether there are any matches, and then does the call you've highlighted here, which will replace all matches.
   - I considered extending the efficient check to identify matches to identify all matches, after which we could manually replace matches with the replacement_text_, but I chose not to as it seemed like it would be difficult to implement/read/maintain, and would potentially be prone to bugs and edge cases. But I could go that route if the performance of `regex_replace` is problematically bad. That being said I think I would prefer to get some performance metrics before doing so
 - Can you link any issues or PRs you can recall where there's discussion about `std::regex` performance issues?

Reply:
> Can you link any issues or PRs you can recall where there's discussion about std::regex performance issues?

[Here's an issue in DLP](https://github.com/solo-io/solo-projects/issues/3957#). Similar situation - regexes are compiled at the time of configuration, but are still extremely slow. There are a few benchmarks linked in the issue that show how... hopeless `std::regex` can be.

Happy to leave this as is for now, but if this issue does bubble up, there's a bit of additional and (hopefully) helpful information on that thread